### ISSUE TITLE
False positive for PERF401 when using assignment expression

### ISSUE THREAD

**Issue by mounirmesselmeni** [astral-sh/ruff#7418]
Hello,

the PERF401 is giving false positive when using [Python's assignment expression](https://peps.python.org/pep-0572/)

Python version: 3.11.5
ruff: 0.0.287

Example code:

```python
results = []
for x in [0, 1, 2, 3, 4]:
    if result := my_func(x):
        results.append(result)
```

Screenshot from VSCode
<img width="708" alt="image" src="https://github.com/astral-sh/ruff/assets/1055731/ed6acf0f-a384-4a95-9f73-3c6391cb2bb4">

PS: Thanks for this great library and amazing work you're doing!

**Comment by charliermarsh** [comment]
Thanks for the kind words :)

I think this snippet _can_ be rewritten as:

```python
results = [
    result for x in [0, 1, 2, 3, 4] if (result := my_func(x))
]
```

Though it's a bit less clear than the originating code, so I'm unsure whether to recommend it or not.

**Comment by mounirmesselmeni** [comment]
Nice one!

Yeah, you can see with the community tbh if Ruff should ignore those use case or not.

For myself now, I can live with this 😄  Thanks again @charliermarsh

**Comment by jamesbraza** [comment]
For posterity/linking, this is a dupe of https://github.com/astral-sh/ruff/issues/6056
